### ISSUE TITLE
Split Registry API from Webhook Processor

### ISSUE THREAD

**Issue by AlexGilleran** [TerriaJS/magda#796]
### Problem description
Right now the Registry API has two responsibilities - providing API access to records in the registry like datasets, distributions etc., AND processing webhooks.

This is a problem because we want to be able to scale the registry API for scale and availability - every other API in Magda has two instances in prod for high availability, but we can't do this for the registry because the webhook processors will step on each other.

It also means that there's no isolation between webhook processing, which isn't that important as we can get around to processing webhooks with delays if we need to, and the Registry API, which is super important because it serves the front-end, and if it goes down then for all intents and purposes the entire system is down. Right now we're seeing the registry api fall over every now and again in production - Kubernetes helpfully detects this and restarts it but sometimes it causes a few seconds of downtimes which is bad, and I suspect that it's the webhook processor's fault.

First step for this would be to separate them while still having them talk to the same database. Second step would be to separate the datastores if possible - ideally we'd decouple the whole idea of event sourcing from the registry completely, so we could use it for other entities in the system.

**Comment by t83714** [comment]
I think it's a very good idea +1

**Comment by aneesha09** [comment]
SPIKE to figure out the best design approach

**Comment by kring** [comment]
Makes sense to me. I was wondering, though: why create a separate row for each event/subscription pair? Instead of the registry's current approach of associating a "last event ID seen" with each subscription? There's a good chance you explained why and I just missed it.

**Comment by AlexGilleran** [comment]
@kring So you can have retries (although I might not have thought this through enough)

Say you have n events and two instances of the same minion consuming the event stream.

- Minion 1 starts up and grabs events 1-10, they're set to `pending` for that subscription
- Minion 2 starts up and grabs events 11-20, they're set to `pending` for that subscription
- Minion 1 gets goes down (maybe the node failed or something)
- Minion 1 starts up again, grabs events 21-30 because those are the next non-pending non-completed events. Events 1-10 are still in state `pending`

* some time passes *

- Minion 1 grabs the next page of events, but enough time has passed that the `lastmodified` timestamp on events 1-10 is greater than the retry threshold. So rather than get the next page of events, it gets 1-10, their lastmodified (maybe this should be called something different) is set to `now()` and they remain in state pending
- Minion 1 doesn't crash this time and completes events 1-10, so they're set to "complete"

Just keeping track of the last event id doesn't allow for that retry. _However_ it also means that events won't necessarily be processed in order. But I think that's OK - it just becomes encumbent upon the consumers of the events not to trust that the data in the event represents the most up-to-date state of the system... which it shouldn't be doing anyway because the event might be 3 months old for all it knows.

**Comment by AlexGilleran** [comment]
> Storing multiple copies of every event makes me nervous, though. There are lots of events.

🤔 yeah. Although hopefully either way (whether keeping a row per event or row per range) it's safe to delete ones that are done. Although that poses a question as how can you be absolutely sure that you're creating a subscriptionevent (or subscriptioneventrange) for each subscription every time there's a new event 😬

**Comment by kring** [comment]
True, it's safe to delete the ones that are done. But still, if there is an event for every subscription, and there are 10 subscriptions, handling a new event (which connectors can generate, I dunno, tens to hundreds of per second, maybe more) requires writing 10 times as much stuff.

In the scheme I outlined above, I don't think it's necessary to change the subscriptioneventrange table for new events. It only needs to be modified when a minion asks for events to process or when it marks them done. This is good because it makes inactive subscriptions (where no one is processing the events) basically free. If a new event is created per subscription, inactive subscriptions are super expensive because their events never get deleted.
